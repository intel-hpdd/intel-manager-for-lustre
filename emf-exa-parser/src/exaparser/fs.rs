// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FsSettings {
    pub backfs: String,
    pub mgs_failback: bool,
    pub host_list: Vec<String>,
    pub ost_mke2fs_opts: Option<Vec<String>>,
    pub total_mdt_count: i64,
    pub tune_ost_mke2fs_opts: Option<Vec<String>>,
    pub mdt_base_device_path: Option<String>,
    pub ost_mount_opts: Option<String>,
    pub mdt_opts: String,
    pub mmp_update_interval: i64,
    pub mgs_size: String,
    pub mdt_parts: Option<i64>,
    pub default_ost_count: i64,
    pub default_mdt_count: i64,
    pub mgs_standalone: bool,
    pub dom_max_file_size: i64,
    pub mds_list: Vec<String>,
    pub log_dir: String,
    pub mdt_list: HashMap<String, Vec<i64>>,
    pub hsm_active: bool,
    pub dom_enabled: bool,
    pub mgt_mount_opts: String,
    pub mdt_failback: bool,
    pub mgs_list: Vec<String>,
    pub oss_list: Vec<String>,
    pub ost_opts: Option<String>,
    pub cur_ost_index: i64,
    pub ost_list: HashMap<String, Vec<i64>>,
    pub ost_lnet_list: HashMap<String, Vec<i64>>,
    pub total_ost_count: i64,
    pub name: String,
    pub ost_device_path: Option<String>,
    pub mgs_internal: bool,
    pub mdt_size: String,
    pub cur_mdt_index: i64,
    pub mgs_device_path: Option<String>,
    pub pools: Vec<String>,
    pub mdt_mount_opts: Option<String>,
    pub mdt_mke2fs_opts: Option<Vec<String>>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]

pub struct PoolSettings {
    pub ost_list: Vec<String>,
    pub name: String,
}
