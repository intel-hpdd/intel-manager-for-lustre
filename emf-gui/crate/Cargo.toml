[package]
authors = ["EMF Team <emf@whamcloud.com>"]
categories = ["wasm", "rust", "tailwindcss", "seed"]
description = "The EMF GUI"
edition = "2018"
license = "MIT"
name = "emf-gui"
readme = "./README.md"
repository = "https://github.com/whamcloud/exascaler-management-framework"
version = "0.3.0"

[lib]
crate-type = ["cdylib"]

[dev-dependencies]
rand_core = "0.5.1"
rand_xoshiro = "0.4.0"
wasm-bindgen-test = "0.3"

[dependencies]
chrono = {version = "0.4", features = ["wasmbind"]}
chrono-humanize = "0.0.11"
futures = "0.3"
gloo-timers = {version = "0.2", features = ["futures"]}
humantime = "2.0"
im = {version = "15", features = ["serde"]}
emf-api-utils = {path = "../../emf-api-utils", version = "0.4"}
emf-graphql-queries = {version = "0.2", path = "../../emf-graphql-queries"}
emf-influx = {path = "../../emf-influx", version = "0.2"}
emf-wire-types = {path = "../../emf-wire-types", version = "0.4"}
js-sys = "0.3"
lazy_static = "1.4.0"
natord = "1.0"
number-formatter = {path = "../../number-formatter", version = "0.2"}
seed = "0.6.0"
serde = {version = "1", features = ['derive']}
serde_json = "1.0"
serde_urlencoded = "0.7.0"
wasm-bindgen = {version = "0.2", features = ["serde-serialize"]}
wasm-bindgen-futures = "0.4"

[dependencies.regex]
default-features = false
features = ["std", "unicode-perl"]
version = "1.3"

[dependencies.web-sys]
features = [
  "DomRect",
  "DomRectReadOnly",
  "Element",
  "EventSource",
  "HtmlDocument",
  "MessageEvent",
  "Navigator",
  "Notification",
  "NotificationOptions",
  "NotificationPermission",
  "ServiceWorkerContainer",
  "ServiceWorkerRegistration",
  "Window",
]
version = "^0.3"

[profile.release]
codegen-units = 1
lto = true
opt-level = 'z'
