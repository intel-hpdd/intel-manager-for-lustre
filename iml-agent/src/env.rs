// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use lazy_static::lazy_static;
use std::{env, path::Path, process::Command};
use url::Url;

/// Checks if the given path exists in the FS
///
/// # Arguments
///
/// * `name` - The path to check
fn path_exists(name: &str) -> bool {
    Path::new(name).exists()
}

/// Gets the environment variable or panics
/// # Arguments
///
/// * `name` - Variable to read from the environment
pub fn get_var(name: &str) -> String {
    env::var(name).unwrap_or_else(|_| panic!("{} environment variable is required.", name))
}

pub fn get_var_else(name: &str, default: &str) -> String {
    env::var(name).unwrap_or_else(|_| default.to_string())
}

lazy_static! {
    // Gets the manager url or panics
    pub static ref MANAGER_URL: Url =
        Url::parse(&get_var("IML_MANAGER_URL")).expect("Could not parse manager url");
}

fn get_private_pem_path() -> String {
    get_var("PRIVATE_PEM_PATH")
}

fn get_cert_path() -> String {
    get_var("CRT_PATH")
}

fn get_pfx_path() -> String {
    get_var("PFX_PATH")
}

fn get_authority_cert_path() -> String {
    get_var("AUTHORITY_CRT_PATH")
}

pub fn sock_dir() -> String {
    get_var("SOCK_DIR")
}

/// Return socket address for a given mailbox
pub fn mailbox_sock(mailbox: &str) -> String {
    format!("{}/postman-{}.sock", sock_dir(), mailbox)
}

lazy_static! {
    // Gets the pfx file.
    // If pfx is not found it will be created.
    pub static ref PFX: Vec<u8> = {
        let private_pem_path = get_private_pem_path();

        if !path_exists(&private_pem_path) {
            panic!("{} does not exist", private_pem_path)
        };

        let cert_path = get_cert_path();

        if !path_exists(&cert_path) {
            panic!("{} does not exist", cert_path)
        }

        let authority_cert_path = get_authority_cert_path();

        let pfx_path = get_pfx_path();

        Command::new("openssl")
            .args(&[
                "pkcs12",
                "-export",
                "-out",
                &pfx_path,
                "-inkey",
                &private_pem_path,
                "-in",
                &cert_path,
                "-certfile",
                &authority_cert_path,
                "-passout",
                "pass:",
            ])
            .status()
            .expect("Error creating pfx");

        std::fs::read(&pfx_path).expect("Could not read pfx")
    };
}
