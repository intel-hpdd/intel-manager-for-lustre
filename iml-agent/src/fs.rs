// Copyright (c) 2019 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use futures::{future::poll_fn, Future, Stream};
use std::{
    io::{BufReader, Write},
    path::Path,
};
use tempfile::NamedTempFile;
use tokio_threadpool::blocking;

/// Given a path, attempts to do an async read to the end of the file.
///
/// # Arguments
///
/// * `p` - The `Path` to a file.
pub fn read_file_to_end<P>(p: P) -> impl Future<Item = Vec<u8>, Error = ImlAgentError>
where
    P: AsRef<Path> + Send + 'static,
{
    tokio::fs::File::open(p).from_err().and_then(|file| {
        tokio::io::read_to_end(file, vec![])
            .map(|(_, d)| d)
            .from_err()
    })
}

/// Given a path, streams the file line by line till EOF
///
/// # Arguments
///
/// * `p` - The `Path` to a file.
pub fn stream_file<P>(p: P) -> impl Stream<Item = String, Error = ImlAgentError>
where
    P: AsRef<Path> + Send + 'static,
{
    tokio::fs::File::open(p)
        .map(|file| tokio::io::lines(BufReader::new(file)))
        .flatten_stream()
        .from_err()
}

/// Given a directory of files,
/// stream each file one at a time till EOF
///
/// # Arguments
///
/// * `p` - The `Path` to a directory.
pub fn stream_dir<P>(p: P) -> impl Stream<Item = String, Error = ImlAgentError>
where
    P: AsRef<Path> + Send + 'static,
{
    tokio::fs::read_dir(p)
        .flatten_stream()
        .map(|d| d.path())
        .map(stream_file)
        .flatten()
}

/// Creates a temporary file and writes some bytes to it.
pub fn write_tempfile(
    contents: Vec<u8>,
) -> impl Future<Item = NamedTempFile, Error = ImlAgentError> {
    poll_fn(move || {
        blocking(|| {
            let mut f = NamedTempFile::new()?;

            f.write_all(contents.as_ref())?;

            Ok(f)
        })
        .map_err(|_| panic!("the threadpool shut down"))
    })
    .and_then(|x| x)
}

#[cfg(test)]
mod tests {
    use super::{stream_dir, stream_file, write_tempfile};
    use crate::agent_error::Result;
    use futures::{Future, Stream as _};
    use std::{fs::File, io::Write as _};
    use tempdir::TempDir;
    use tempfile::NamedTempFile;

    fn run<R: Send + 'static, E: Send + 'static>(
        fut: impl Future<Item = R, Error = E> + Send + 'static,
    ) -> std::result::Result<R, E> {
        tokio::runtime::Runtime::new().unwrap().block_on_all(fut)
    }

    #[test]
    fn test_write_tempfile() {
        let fut = write_tempfile(b"foobar".to_vec());

        let file = run(fut).unwrap();

        let s = std::fs::read_to_string(file.path()).unwrap();

        assert_eq!(s, "foobar");
    }

    #[test]
    fn test_stream_file() -> Result<()> {
        let mut f = NamedTempFile::new()?;

        f.write_all(b"some\nawesome\nfile")?;

        let path = f.path().to_path_buf();

        let fut = stream_file(path).collect();

        let result = run(fut)?;

        assert_eq!(result, vec!["some", "awesome", "file"]);

        Ok(())
    }

    #[test]
    fn test_stream_dir() -> Result<()> {
        let tmp_dir = TempDir::new("test_stream_dir")?;

        for i in 1..=5 {
            let file_path = tmp_dir.path().join(format!("test_file{}.txt", i));
            let mut tmp_file = File::create(file_path)?;
            writeln!(tmp_file, "file{}\nwas{}\nhere{}", i, i, i)?;
        }

        let fut = stream_dir(tmp_dir.path().to_path_buf()).chunks(3).collect();

        let mut result = run(fut)?;
        result.sort();

        assert_eq!(
            result,
            vec![
                vec!["file1", "was1", "here1"],
                vec!["file2", "was2", "here2"],
                vec!["file3", "was3", "here3"],
                vec!["file4", "was4", "here4"],
                vec!["file5", "was5", "here5"]
            ]
        );

        Ok(())
    }
}
