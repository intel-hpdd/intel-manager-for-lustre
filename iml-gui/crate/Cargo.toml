[package]
authors = ["IML Team <iml@whamcloud.com>"]
categories = ["wasm", "rust", "tailwindcss", "seed"]
description = "The IML GUI"
edition = "2018"
license = "MIT"
name = "iml-gui"
readme = "./README.md"
repository = "https://github.com/whamcloud/integrated-manager-for-lustre"
version = "0.3.0"

[lib]
crate-type = ["cdylib"]

[dev-dependencies]
rand_core = "0.5.1"
rand_xoshiro = "0.4.0"
wasm-bindgen-test = "0.3"

[dependencies]
chrono = {version = "0.4", features = ["wasmbind"]}
chrono-humanize = "0.0.11"
futures = "0.3"
gloo-timers = {version = "0.2", features = ["futures"]}
humantime = "2.0"
im = {version = "15", features = ["serde"]}
iml-api-utils = {path = "../../iml-api-utils", version = "0.4"}
iml-graphql-queries = {version = "0.2", path = "../../iml-graphql-queries"}
iml-influx = {path = "../../iml-influx", version = "0.2"}
iml-wire-types = {path = "../../iml-wire-types", version = "0.4"}
js-sys = "0.3"
lazy_static = "1.4.0"
natord = "1.0"
number-formatter = {path = "../../number-formatter", version = "0.2"}
seed = "0.6.0"
serde = {version = "1", features = ['derive']}
serde_json = "1.0"
serde_urlencoded = "0.7.0"
wasm-bindgen = {version = "0.2", features = ["serde-serialize"]}
wasm-bindgen-futures = "0.4"

[dependencies.regex]
default-features = false
features = ["std", "unicode-perl"]
version = "1.3"

[dependencies.web-sys]
features = [
  "DomRect",
  "DomRectReadOnly",
  "Element",
  "EventSource",
  "HtmlDocument",
  "MessageEvent",
  "Navigator",
  "Notification",
  "NotificationOptions",
  "NotificationPermission",
  "ServiceWorkerContainer",
  "ServiceWorkerRegistration",
  "Window",
]
version = "^0.3"

[profile.release]
codegen-units = 1
lto = true
opt-level = 'z'
