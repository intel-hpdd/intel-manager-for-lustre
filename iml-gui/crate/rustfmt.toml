# https://github.com/rust-lang/rustfmt/blob/master/Configurations.md

edition = "2018"
newline_style = "Unix"
use_field_init_shorthand = true
use_try_shorthand = true

max_width = 120
