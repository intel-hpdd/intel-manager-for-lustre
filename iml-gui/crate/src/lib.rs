// @TODO: uncomment once https://github.com/rust-lang/rust/issues/54726 stable
//#![rustfmt::skip::macros(class)]

#![allow(clippy::used_underscore_binding)]
#![allow(clippy::non_ascii_literal)]
#![allow(clippy::enum_glob_use)]

mod generated;
mod page;

use generated::css_classes::C;
use iml_wire_types::warp_drive;
use js_sys::Function;
use seed::{events::Listener, prelude::*, *};
use std::{cmp, collections::HashMap, mem};
use wasm_bindgen::JsCast;
use web_sys::{EventSource, MessageEvent};
use Visibility::*;

const TITLE_SUFFIX: &str = "IML";
const USER_AGENT_FOR_PRERENDERING: &str = "ReactSnap";
const STATIC_PATH: &str = "static";
const SLIDER_WIDTH_PX: u32 = 5;
const MAX_SIDE_PERCENTAGE: f32 = 35f32;

#[derive(Clone, Copy, Eq, PartialEq)]
pub enum Visibility {
    Visible,
    Hidden,
}

impl Visibility {
    pub fn toggle(&mut self) {
        *self = match self {
            Visible => Hidden,
            Hidden => Visible,
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum WatchState {
    Watching,
    Open,
    Close,
}

impl Default for WatchState {
    fn default() -> Self {
        WatchState::Close
    }
}

impl WatchState {
    pub fn is_open(self) -> bool {
        match self {
            WatchState::Open => true,
            _ => false,
        }
    }
    pub fn is_watching(self) -> bool {
        match self {
            WatchState::Watching => true,
            _ => false,
        }
    }
    pub fn should_update(self) -> bool {
        self.is_watching() || self.is_open()
    }
    pub fn update(&mut self) {
        match self {
            WatchState::Close => {
                mem::replace(self, WatchState::Watching);
            }
            WatchState::Watching => {
                mem::replace(self, WatchState::Open);
            }
            WatchState::Open => {
                mem::replace(self, WatchState::Close);
            }
        }
    }
}

// ------ ------
//     Model
// ------ ------

pub struct Model {
    pub page: Page,
    pub menu_visibility: Visibility,
    pub in_prerendering: bool,
    pub config_menu_state: WatchState,
    pub track_slider: bool,
    pub side_width_percentage: f32,
    pub records: warp_drive::Cache,
    pub locks: warp_drive::Locks,
}

#[derive(Clone, Copy, Eq, PartialEq)]
pub enum Page {
    Dashboard,
    Home,
    About,
    NotFound,
    Jobstats,
}

impl Page {
    pub fn to_href(self) -> &'static str {
        match self {
            Self::Home => "/",
            Self::Dashboard => "/dashboard",
            Self::Jobstats => "/jobstats",
            Self::About => "/about",
            Self::NotFound => "/404",
        }
    }
}

impl ToString for Page {
    fn to_string(&self) -> String {
        match self {
            Self::Dashboard => "Dashboard".into(),
            Self::Home => "Home".into(),
            Self::About => "About".into(),
            Self::NotFound => "404".into(),
            Self::Jobstats => "Jobstats".into(),
        }
    }
}

impl From<Url> for Page {
    fn from(url: Url) -> Self {
        match url.path.first().map(String::as_str) {
            None | Some("") => Self::Home,
            Some("dashboard") => Self::Dashboard,
            Some("about") => Self::About,
            Some("jobstats") => Self::Jobstats,
            _ => Self::NotFound,
        }
    }
}

pub fn register_eventsource_handle<T, F>(
    es_cb_setter: fn(&EventSource, Option<&Function>),
    msg: F,
    ws: &EventSource,
    orders: &mut impl Orders<Msg>,
) where
    T: wasm_bindgen::convert::FromWasmAbi + 'static,
    F: Fn(T) -> Msg + 'static,
{
    let (app, msg_mapper) = (orders.clone_app(), orders.msg_mapper());

    let closure = Closure::new(move |data| {
        app.update(msg_mapper(msg(data)));
    });

    es_cb_setter(ws, Some(closure.as_ref().unchecked_ref()));
    closure.forget();
}

// ------ ------
//     Init
// ------ ------

pub fn init(url: Url, orders: &mut impl Orders<Msg>) -> Init<Model> {
    // @TODO: Seed can't hydrate prerendered html (yet).
    // https://github.com/David-OConnor/seed/issues/223
    if let Some(mount_point_element) = document().get_element_by_id("app") {
        mount_point_element.set_inner_html("");
    }

    let es = EventSource::new("https://localhost:7444/messaging").unwrap();

    register_eventsource_handle(
        EventSource::set_onopen,
        Msg::EventSourceConnect,
        &es,
        orders,
    );

    register_eventsource_handle(
        EventSource::set_onmessage,
        Msg::EventSourceMessage,
        &es,
        orders,
    );

    register_eventsource_handle(
        EventSource::set_onerror,
        Msg::EventSourceError,
        &es,
        orders,
    );

    orders.send_msg(Msg::UpdatePageTitle);

    Init::new(Model {
        page: url.into(),
        menu_visibility: Visible,
        in_prerendering: is_in_prerendering(),
        config_menu_state: WatchState::default(),
        track_slider: false,
        side_width_percentage: 20f32,
        records: warp_drive::Cache::default(),
        locks: HashMap::new(),
    })
}

fn is_in_prerendering() -> bool {
    let user_agent =
        window().navigator().user_agent().expect("cannot get user agent");

    user_agent == USER_AGENT_FOR_PRERENDERING
}

// ------ ------
//    Routes
// ------ ------

pub fn routes(url: Url) -> Option<Msg> {
    // Urls which start with `static` are files => treat them as external links.
    if url.path.starts_with(&[STATIC_PATH.into()]) {
        return None;
    }
    Some(Msg::RouteChanged(url))
}

// ------ ------
//    Update
// ------ ------

#[derive(Clone)]
pub enum Msg {
    RouteChanged(Url),
    UpdatePageTitle,
    ToggleMenu,
    ConfigMenuState,
    HideMenu,
    StartSliderTracking,
    StopSliderTracking,
    SliderX(i32, f64),
    EventSourceConnect(JsValue),
    EventSourceMessage(MessageEvent),
    EventSourceError(JsValue),
    Records(warp_drive::Cache),
    RecordChange(warp_drive::RecordChange),
    Locks(warp_drive::Locks),
    WindowClick,
}

pub fn update(msg: Msg, model: &mut Model, orders: &mut impl Orders<Msg>) {
    match msg {
        Msg::RouteChanged(url) => {
            model.page = url.into();
            orders.send_msg(Msg::UpdatePageTitle);
        }
        Msg::UpdatePageTitle => {
            let title = match model.page {
                Page::Home => TITLE_SUFFIX.to_owned(),
                Page::Dashboard => format!("Dashboard - {}", TITLE_SUFFIX),
                Page::Jobstats => format!("Jobstats - {}", TITLE_SUFFIX),
                Page::About => format!("About - {}", TITLE_SUFFIX),
                Page::NotFound => format!("404 - {}", TITLE_SUFFIX),
            };
            document().set_title(&title);
        }
        Msg::EventSourceConnect(_) => {
            log("EventSource connected.");
        }
        Msg::EventSourceMessage(msg) => {
            let txt = msg.data().as_string().unwrap();

            let msg: warp_drive::Message = serde_json::from_str(&txt).unwrap();

            let msg = match msg {
                warp_drive::Message::Locks(locks) => Msg::Locks(locks),
                warp_drive::Message::Records(records) => Msg::Records(records),
                warp_drive::Message::RecordChange(record_change) => {
                    Msg::RecordChange(record_change)
                }
            };

            orders.send_msg(msg);
        }
        Msg::EventSourceError(_) => {
            log("EventSource error.");
        }
        Msg::Records(records) => {
            model.records = records;
        }
        Msg::RecordChange(record_change) => match record_change {
            warp_drive::RecordChange::Update(record) => match record {
                warp_drive::Record::ActiveAlert(x) => {
                    model.records.active_alert.insert(x.id, x);
                }
                warp_drive::Record::Filesystem(x) => {
                    model.records.filesystem.insert(x.id, x);
                }
                warp_drive::Record::Host(x) => {
                    model.records.host.insert(x.id, x);
                }
                warp_drive::Record::ManagedTargetMount(x) => {
                    model.records.managed_target_mount.insert(x.id, x);
                }
                warp_drive::Record::StratagemConfig(x) => {
                    model.records.stratagem_config.insert(x.id, x);
                }
                warp_drive::Record::Target(x) => {
                    model.records.target.insert(x.id, x);
                }
                warp_drive::Record::Volume(x) => {
                    model.records.volume.insert(x.id, x);
                }
                warp_drive::Record::VolumeNode(x) => {
                    model.records.volume_node.insert(x.id, x);
                }
                warp_drive::Record::LnetConfiguration(x) => {
                    model.records.lnet_configuration.insert(x.id, x);
                }
            },
            warp_drive::RecordChange::Delete(record_id) => match record_id {
                warp_drive::RecordId::ActiveAlert(x) => {
                    model.records.active_alert.remove(&x);
                }
                warp_drive::RecordId::Filesystem(x) => {
                    model.records.filesystem.remove(&x);
                }
                warp_drive::RecordId::Host(x) => {
                    model.records.host.remove(&x);
                }
                warp_drive::RecordId::ManagedTargetMount(x) => {
                    model.records.managed_target_mount.remove(&x);
                }
                warp_drive::RecordId::StratagemConfig(x) => {
                    model.records.stratagem_config.remove(&x);
                }
                warp_drive::RecordId::Target(x) => {
                    model.records.target.remove(&x);
                }
                warp_drive::RecordId::Volume(x) => {
                    model.records.volume.remove(&x);
                }
                warp_drive::RecordId::VolumeNode(x) => {
                    model.records.volume_node.remove(&x);
                }
                warp_drive::RecordId::LnetConfiguration(x) => {
                    model.records.lnet_configuration.remove(&x);
                }
            },
        },
        Msg::Locks(locks) => {
            model.locks = locks;
        }
        Msg::ToggleMenu => model.menu_visibility.toggle(),
        Msg::ConfigMenuState => {
            model.config_menu_state.update();
        }
        Msg::HideMenu => {
            model.menu_visibility = Hidden;
        }
        Msg::StartSliderTracking => {
            model.track_slider = true;
        }
        Msg::StopSliderTracking => {
            model.track_slider = false;
        }
        Msg::SliderX(x_position, page_width) => {
            let overlay_width_px = page_width as u32 - SLIDER_WIDTH_PX;

            let x_position = cmp::max(0, x_position) as u32;

            let side_width_percentage: f32 =
                (x_position as f32 / overlay_width_px as f32) * 100_f32;

            model.side_width_percentage =
                if MAX_SIDE_PERCENTAGE <= side_width_percentage {
                    MAX_SIDE_PERCENTAGE
                } else {
                    side_width_percentage
                };
        }
        Msg::WindowClick => {
            if model.config_menu_state.should_update() {
                model.config_menu_state.update();
            }
        }
    }
}

// ------ ------
//     View
// ------ ------

// Notes:
// - \u{00A0} is the non-breaking space
//   - https://codepoints.net/U+00A0
//
// - "▶\u{fe0e}" - \u{fe0e} is the variation selector, it prevents ▶ to change to emoji in some browsers
//   - https://codepoints.net/U+FE0E

pub fn view(model: &Model) -> impl View<Msg> {
    // @TODO: Setup `prerendered` properly once https://github.com/David-OConnor/seed/issues/223 is resolved
    let prerendered = true;

    div![
        class![
            C.fade_in => !prerendered,
            C.min_h_screen,
            C.flex,
            C.flex_col,
            C.select_none => model.track_slider
        ],
        // slider overlay
        if model.track_slider {
            div![
                class![
                    C.w_full,
                    C.h_full,
                    C.fixed,
                    C.top_0,
                    C.cursor_ew_resize,
                ],
                style! { St::ZIndex => 9999 },
                mouse_ev(Ev::MouseMove, |ev| {
                    let target = ev.target().unwrap();
                    let el = seed::to_html_el(&target);

                    let rect = el.get_bounding_client_rect();

                    Msg::SliderX(ev.client_x(), rect.width())
                }),
            ]
        } else {
            empty![]
        },
        page::partial::header::view(model).els(),
        // panel container
        div![
            class![
                C.flex,
                C.flex_wrap,
                C.flex_col,
                C.lg__flex_row,
                C.flex_grow
            ],
            // side panel
            div![
                class![
                    C.flex_grow_0,
                    C.flex_shrink_0,
                    C.overflow_x_hidden,
                    C.overflow_y_auto,
                    C.whitespace_no_wrap,
                    C.bg_blue_900,
                    C.border_r_2,
                    C.border_gray_800,
                    C.lg__h_main_content,
                ],
                style! { St::FlexBasis => percent(model.side_width_percentage) },
            ],
            // slider panel
            div![
                class![
                    C.flex_grow_0,
                    C.flex_shrink_0,
                    C.cursor_ew_resize,
                    C.bg_gray_500
                    C.hover__bg_teal_400,
                    C.bg_teal_400 => model.track_slider,
                    C.relative,
                    C.lg__block,
                    C.lg__h_main_content,
                    C.hidden
                ],
                simple_ev(Ev::MouseDown, Msg::StartSliderTracking),
                style! {
                    St::FlexBasis => px(SLIDER_WIDTH_PX),
                },
                div![
                    class![C.absolute, C.rounded],
                    style! {
                        St::BackgroundColor => "inherit",
                        St::Height => px(64),
                        St::Width => px(18),
                        St::Top => "calc(50% - 32px)",
                        St::Left => px(-7.5),
                    }
                ]
            ],
            // main panel
            div![
                class![
                    C.flex,
                    C.flex_col,
                    C.flex_grow,
                    C.flex_shrink_0,
                    C.bg_gray_200,
                    C.lg__w_0,
                    C.lg__h_main_content,
                ],
                // main content
                div![
                    class![
                        C.flex_grow,
                        C.overflow_x_auto,
                        C.overflow_y_auto,
                        C.p_6
                    ],
                    match model.page {
                        Page::Home => page::home::view(&model).els(),
                        Page::Dashboard => page::dashboard::view(&model).els(),
                        Page::About => page::about::view(&model).els(),
                        Page::Jobstats => page::jobstats::view(&model).els(),
                        Page::NotFound => page::not_found::view(&model).els(),
                    },
                ],
                page::partial::footer::view().els(),
            ]
        ],
    ]
}

pub fn asset_path(asset: &str) -> String {
    format!("{}/{}", STATIC_PATH, asset)
}

// ------ ------
// Window Events
// ------ ------

pub fn window_events(model: &Model) -> Vec<Listener<Msg>> {
    let mut xs = vec![simple_ev(Ev::Click, Msg::WindowClick)];

    if model.track_slider {
        xs.push(simple_ev(Ev::MouseUp, Msg::StopSliderTracking));
    }

    xs
}

// ------ ------
//     Start
// ------ ------

#[wasm_bindgen(start)]
pub fn run() {
    log!("Starting app...");

    App::build(init, update, view)
        .routes(routes)
        .window_events(window_events)
        .finish()
        .run();

    log!("App started.");
}
