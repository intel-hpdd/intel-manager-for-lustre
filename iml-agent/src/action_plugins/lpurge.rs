// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{agent_error::ImlAgentError, env};
use futures::future::TryFutureExt;
use std::{collections::HashMap, path::PathBuf};
use tokio::fs;

#[derive(serde::Deserialize, structopt::StructOpt, Debug)]
pub struct Config {
    #[structopt(long)]
    /// Filesystem Name
    fs: String,
    #[structopt(long)]
    /// Ost Index
    ost: u32,
    #[structopt(long)]
    /// OST pool name
    pool: String,
    #[structopt(long)]
    freelo: u8,
    #[structopt(long)]
    freehi: u8,
    #[structopt(long)]
    mailbox: String,
}

impl Config {
    fn generate_unit(&self, socket: String) -> String {
        format!(
            "\
             device={fs}-OST{ost:04x}\n\
             freehi={freehi}\n\
             freelo={freelo}\n\
             iml_socket={socket}\n\
             max_jobs=0\n\
             pool={fs}.{pool}\n\
             ",
            freehi = self.freehi,
            freelo = self.freelo,
            socket = socket,
            ost = self.ost,
            fs = self.fs,
            pool = self.pool,
        )
    }
}

fn expand_path_fmt(path_fmt: &str, c: &Config) -> strfmt::Result<String> {
    let mut vars = HashMap::new();
    let ost = format!("OST{:04x}", c.ost);
    let freehi = c.freehi.to_string();
    let freelo = c.freelo.to_string();

    vars.insert("fs".to_string(), &c.fs);
    vars.insert("ost".to_string(), &ost);
    vars.insert("pool".to_string(), &c.pool);
    vars.insert("freehi".to_string(), &freehi);
    vars.insert("freelo".to_string(), &freelo);
    vars.insert("mailbox".to_string(), &c.mailbox);
    strfmt::strfmt(&path_fmt, &vars)
}

pub async fn create_lpurge_conf(c: Config) -> Result<(), ImlAgentError> {
    let path_fmt = env::get_var("LPURGE_CONF_PATH");
    let socket = env::mailbox_sock(&c.mailbox);
    let file = conf_name(&c, &path_fmt).await?;
    write(file, &c, socket).err_into().await
}

async fn conf_name(c: &Config, path_fmt: &str) -> Result<PathBuf, ImlAgentError> {
    let path = PathBuf::from(expand_path_fmt(path_fmt, c)?);
    Ok(path)
}

async fn write(file: PathBuf, c: &Config, socket: String) -> std::io::Result<()> {
    if let Some(parent) = file.parent() {
        fs::create_dir_all(&parent).await?;
    }

    let cnt = c.generate_unit(socket);

    fs::write(file, cnt.as_bytes()).await
}

#[cfg(test)]
mod lpurge_conf_tests {
    use super::*;
    use insta::assert_display_snapshot;
    use std::env;
    use tempfile::tempdir;

    #[tokio::test]
    async fn works() {
        let cfg = Config {
            fs: "lima".to_string(),
            pool: "santiago".to_string(),
            ost: 16,
            freehi: 123,
            freelo: 60,
            mailbox: "foobar".to_string(),
        };

        let dir = tempdir().expect("could not create temp file");
        let file = dir.path().join("config");
        let file2 = file.clone();

        write(file, &cfg, "foobar".to_string())
            .await
            .expect("could not write");

        let cnt = String::from_utf8(std::fs::read(&file2).expect("could not read file")).unwrap();

        assert_display_snapshot!(cnt);
    }

    #[tokio::test]
    async fn config_name() {
        let cfg = Config {
            fs: "lima".to_string(),
            pool: "santiago".to_string(),
            ost: 16,
            freehi: 123,
            freelo: 60,
            mailbox: "foobar".to_string(),
        };
        let file = conf_name(&cfg, "/etc/lpurge/{fs}/{ost}-{pool}.conf")
            .await
            .expect("name could not be created");

        assert_eq!(
            file,
            PathBuf::from("/etc/lpurge/lima/OST0010-santiago.conf")
        );
    }
}
