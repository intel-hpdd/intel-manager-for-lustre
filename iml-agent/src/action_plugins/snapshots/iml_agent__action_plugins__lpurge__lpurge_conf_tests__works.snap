---
source: iml-agent/src/action_plugins/lpurge.rs
expression: cnt
---
device=lima-OST0010
freehi=123
freelo=60
iml_socket=foobar
max_jobs=0
pool=lima.santiago

